llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -438,6 +450,161 @@ bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
   return tryToDuplicateLoadUse(LoadUses, NonLoadUses, MRI, TII);
 }
 
+/// This optimization tries to propagate GlobalValues as MMO of load operations
+/// when they are missing.
+bool findGlobalValues(const MachineInstr *MI,
+                      SmallPtrSet<const Value *, 4> &GVSet,
+                      SmallPtrSet<const MachineInstr *, 4> &VisitedInstrs,
+                      MachineRegisterInfo &MRI, unsigned SearchLimit) {
+
+  // Exhausted the search.
+  if (VisitedInstrs.size() == SearchLimit)
+    return false;
+
+  // Skip copies.
+  while (MI->isCopy() && MI->getOperand(1).getReg().isVirtual())
+    MI = MRI.getVRegDef(MI->getOperand(1).getReg());
+
+  // Loaded values cannot be tracked as Globals, but they can
+  // be part of a final address calculation.
+  if (MI->mayLoad())
+    return true;
+
+  if (VisitedInstrs.find(MI) != VisitedInstrs.end())
+    return true;
+
+  VisitedInstrs.insert(MI);
+  bool Success = true;
+  for (auto MO : MI->uses()) {
+    if (MO.isGlobal()) {
+      GVSet.insert(MO.getGlobal());
+    } else if (MO.isReg() && MO.getReg().isVirtual()) {
+      if (MO.getReg().isVirtual()) {
+        const Register Reg = MO.getReg();
+        if (const MachineInstr *RegDef = MRI.getVRegDef(Reg))
+          Success &=
+              findGlobalValues(RegDef, GVSet, VisitedInstrs, MRI, SearchLimit);
+      } else {
+        // Physical registers can be anything (parameters).
+        return false;
+      }
+    }
+  }
+  return Success;
+}
+
+using Opcodes = AIEBaseInstrInfo::AbstractVecOp::AbstractOpcode;
+using AbstractVecOp = AIEBaseInstrInfo::AbstractVecOp;
+using OptionalVecOp = std::optional<AbstractVecOp>;
+
+/// This optimization tries to propagate GlobalValues as MMO of load operations
+/// when they are missing. This specific implementation considers only the
+/// following pattern:
+/// VADD (VSEL (VBCST(GlobalValue1), VBCST(GlobalValue2), Mask) Value)
+bool findBroadcastedGlobalValues(const MachineInstr *MI,
+                                 SmallPtrSet<const Value *, 4> &GVSet,
+                                 MachineRegisterInfo &MRI) {
+
+  const TargetSubtargetInfo &ST = MI->getParent()->getParent()->getSubtarget();
+  const AIEBaseInstrInfo *TII =
+      static_cast<const AIEBaseInstrInfo *>(ST.getInstrInfo());
+
+  // Skip subvector copy.
+  if (MI->isCopy()) {
+    Register Src = MI->getOperand(1).getReg();
+    if (Src.isVirtual())
+      MI = MRI.getVRegDef(Src);
+    else
+      // We cannot proceed with physical registers.
+      return false;
+  }
+
+  auto ParseByReg = [&](Register Reg, Opcodes Opcode) -> OptionalVecOp {
+    OptionalVecOp VOp = TII->parseTargetVectorOp(*MRI.getVRegDef(Reg));
+    if (VOp && VOp->Opcode == Opcode)
+      return VOp;
+    return {};
+  };
+
+  const OptionalVecOp VAddOp = TII->parseTargetVectorOp(*MI);
+
+  // We start with an ADD.
+  if (!VAddOp || VAddOp->Opcode != Opcodes::ADD)
+    return false;
```

Comment:
Nit: I would actually make the pattern a bit more generic. I think we can "bypass" `VADD`s as long as one of their operands is a vector of addresses. I define "vector of addresses" as: an arbitrary number of VSEL.32 that all terminate with VBCST.32 sources. And all those sources are base addresses we can track.

That being said, if that pattern works, I guess it's okay, and we can refine it as we see fit. I'd just prefer to make it a bit more generic if that is not too much work.

Reply:
Sorry I think what I wrote wasn't very clear. I actually meant to make the pattern "broader". I think there could be any number (including 0) of `VADD` instructions to start with. The only thing that really matters is that one of the added vectors is a "vector of pointers".

I would be happy to have a function like `bool isVectorOfPointers(SmallVEctor<Value*> &Pointers)`, that explicitly looks for `VSEL.32` and `VBCST.32` operations.